Add Format option to Crop node

HughMacdonald opened the pull request:
This PR adds an extra option to the Crop node to specify a Format to crop to.
It also adds a checkbox that allows the previous DisplayWindow to be centered in the new Format.

I have a couple of questions on this:

* I have renamed the enum in C++ to add Area to the end of each name, as just Format was clashing. I haven't added this to the Python bindings, as the Python enum values are referenced using GafferImage.Crop.AreaSource.Format, while the C++ enum values are referenced directly.
* There is an option for a user to select Format and then select Custom from the Format menu. Can I suggest a UI option to control whether Custom is available in the Format drop-down list.

Other than that, all feedback welcome.

andrewkaufman commented:
I think it'd be nicer to keep the c++ and python enum values in sync. Where was `Format` clashing? In the cpp? If so, I think you could get around that by differentiating with explicit namespaces (e.g. `GafferImage::Format` and `Crop::Format`). Or was the header complaining?

In general, I find it annoying that c++ enums aren't namespaced by enum name... but I guess c++11 fixes that for us if we can ever get there across the board...

HughMacdonald commented:
There you go - that should do it.

It was previous complaining about a Format that needed to be changed to GafferImage::Format.

johnhaddon commented:
Hi Hugh, thanks for this one.

I have a few fairly minor comments, but I'd like to discuss and possibly address them before merging :

- I think the format settings should appear below the area settings - area mode is still the default so the area plug should be presented first. Can we update the order of plugs in the node, in the header, in the UI, and in the presets menu please, so that format comes below area? The affect/reset options can still come below the format, since they're about something that happens after the area is chosen.
- Can we just call the plugs "format" and "centerFormat"? I find the "area" prefix doesn't really clarify anything for me - if anything it confuses me by associating it with the "area" plug.
- I find the center mode confusing, particularly in its interaction with the resetOrigin setting. When cropping to a format, I would expect the output format to always correspond to the format I chose, but with resetOrigin off, I end up with a weird negative origin. Would it make more sense to present centering as "centers the _input_ image within the _output_ format"? That would mean the output format would be unaffected by centering, and resetOrigin would then only come into play when the area has a non-zero origin (as it does for the other modes).

HughMacdonald commented:
Will address tomorrow (currently stuck in bed with a trapped nerve in my
shoulder...)

Comments/counter questions below:

On 23 Nov 2015, at 22:26, John Haddon <notifications@github.com> wrote:

Hi Hugh, thanks for this one.

I have a few fairly minor comments, but I'd like to discuss and possibly
address them before merging :

   - I think the format settings should appear below the area settings -
   area mode is still the default so the area plug should be presented first.
   Can we update the order of plugs in the node, in the header, in the UI, and
   in the presets menu please, so that format comes below area? The
   affect/reset options can still come below the format, since they're about
   something that happens after the area is chosen.

That was my initial thought, but I've always felt that "Custom" should go
at the bottom of any list. I also made the plugs below in the same order as
the options in the list.

Can I propose a naming change (that will break backwards compatibility)

Change "Area From" to "Crop From"
Change "Custom" to "Area", and move to top of list.
Move "Format" to second in list.

The other issue, which could be controlled with grouping, was that having
the "Format Center" checkbox right next to "Affect Data Window" looked like
it belonged with the checkboxes below rather than the drop-down list above.


   - Can we just call the plugs "format" and "centerFormat"? I find the
   "area" prefix doesn't really clarify anything for me - if anything it
   confuses me by associating it with the "area" plug.

I was considering the result of all of that section to be the "area", so
the "area" plug should really have been called "Area Custom". If we rename
"Area From" as above, then yes, I think this makes sense.


   - I find the center mode confusing, particularly in its interaction with
   the resetOrigin setting. When cropping to a format, I would expect the
   output format to always correspond to the format I chose, but with
   resetOrigin off, I end up with a weird negative origin. Would it make more
   sense to present centering as "centers the *input* image within the
   *output* format"? That would mean the output format would be unaffected
   by centering, and resetOrigin would then only come into play when the area
   has a non-zero origin (as it does for the other modes).

Yup, that makes perfect sense. I'll have to remind myself of the inner
workings. Do you think that having "Center" be available for any crop type
rather than just format makes sense?

Thinking a bit further, is there any point to allowing cantering when we
aren't affecting the display window? I'm not sure there is. I'm thinking
that it's only applicable when the display window is being changed.



—
Reply to this email directly or view it on GitHub
<https://github.com/ImageEngine/gaffer/pull/1570#issuecomment-159087596>.

johnhaddon commented:
Hi Hugh, hope the shoulder's on the mend. Sorry for the delay in replying to this one - I actually composed a reply yesterday but it appears not to have made it through the ether. 

> I've always felt that "Custom" should go at the bottom of any list. I also made the plugs below in the same order as the options in the list. Can I propose a naming change (that will break backwards compatibility) Change "Area From" to "Crop From" Change "Custom" to "Area", and move to top of list. Move "Format" to second in list.

I like the "Custom" -> "Area" name change, I've felt for a while now that the name didn't fit. With that change in place and with the order you suggested I'd be happy enough I think - no need to rename "areaSource" to my mind.

> The other issue, which could be controlled with grouping, was that having the "Format Center" checkbox right next to "Affect Data Window" looked like it belonged with the checkboxes below rather than the drop-down list above.

I'd suggest we just add a simple divider between the "area source" and "affects" plugs then. You can do that with a `"divider", True` metadata entry.

>- Can we just call the plugs "format" and "centerFormat"? I find the "area" prefix doesn't really clarify anything for me - if anything it confuses me by associating it with the "area" plug. I was considering the result of all of that section to be the "area", so the "area" plug should really have been called "Area Custom". If we rename "Area From" as above, then yes, I think this makes sense. 

I think this makes sense even without renaming "Area From". It's a pretty small node, so it's not like we need to namespace every single plug with some prefix to clarify its purpose. I think a sensible ordering, activators, and a discreet divider should make everything pretty clear.

> - I find the center mode confusing, particularly in its interaction with the resetOrigin setting. Would it make more sense to present centering as "centers the *input* image within the *output* format"?
Yup, that makes perfect sense.

Great.

> I'll have to remind myself of the inner workings. Do you think that having "Center" be available for any crop type rather than just format makes sense?

It doesn't feel like it would be useful to me. The other area sources (area, display window, data window) are already tied intrinsically to the image itself, so I don't see a shift being useful there. I suggest we keep it simple and have it only as an option for the format mode.

> Thinking a bit further, is there any point to allowing cantering when we aren't affecting the display window? I'm not sure there is. I'm thinking that it's only applicable when the display window is being changed.

That sounds right to me. Additionally I find it hard to imagine using the "Format" source without applying it to the display window, so I think that's another argument for centering only applying to the Format mode.

This brings up an awkward problem I hadn't considered till now though. What if the new format has a different pixel aspect to the input format? I suspect we'd then want to resample to avoid squashing, which maybe brings us full circle back to this being a mode in Resize rather than Crop?

HughMacdonald commented:
I think I've addressed all of your comments there...

With regards to your last question, I think that this node should very specifically never do any sampling. I would be okay with it keeping the current functionality, and setting the pixel aspect ratio in the output format. Any squashing of pixels should be done in Resize, but I wouldn't suggest a "None" mode in Resize that does affect the pixels for this reason.

HughMacdonald commented:
Odd - the failed test is coming from something completely different. I can't re-launch the build, though. My own Travis build succeeded.

HughMacdonald commented:
It looks like this is all good to me. Would be good to know if you guys want any changes to it.

HughMacdonald commented:
I've reordered the enum values. Let me know if you'd like a compatibility for Custom -> Area and I can add that too.

andrewkaufman commented:
Might as well since its so simple

HughMacdonald commented:
Cool stuff. I'll drop that in first thing tomorrow then.

HughMacdonald commented:
Added compatibility code

johnhaddon commented:
Thanks Hugh. I've fixed up the activator for the formatCenter plug and merged manually - let me know if my fix doesn't leave things as you'd intended…
Cheers…
John

HughMacdonald commented:
Oops - yes - that looks right. I was mid-way through getting that working at one point, and appear to have forgotten about it.

I take it what I was trying with two activators doesn't work. And a typo would be why my other activator wasn't doing its job.

johnhaddon commented:
> I take it what I was trying with two activators doesn't work. And a typo would be why my other activator wasn't doing its job.

Yep - exactly. You can only have one metadata value for each key, so your second declaration was overwriting the first. It's definitely valid to overwrite a pre-existing metadata value, but it might be nice to have `registerNode()` check that you're not doing that from the exact same call, since that would most likely indicate a mistake.
